backend/app/verification.py

```diff
@@ -447,13 +448,54 @@ def _verify_by_github(username: str, account) -> bool:
         raise HTTPException(status_code=500, detail=ErrorDetail.PROVIDER_ERROR)
 
 
-def _verify_by_account(username: str, account, caseSensitive: bool):
+def _verify_by_gitlab(username: str, account, model, url) -> bool:
+    account = model.by_user(sqldb, account)
+
     if account is None:
         raise HTTPException(status_code=401, detail=ErrorDetail.NOT_LOGGED_IN)
-    elif account.login.lower() != username.lower() or (
-        caseSensitive and account.login != username
-    ):
-        raise HTTPException(status_code=403, detail=ErrorDetail.USERNAME_DOES_NOT_MATCH)
+
+    # If the username matches, return immediately. We use .lower() because GitLab usernames are case insensitive.
+    if account.login.lower() == username.lower():
+        return False
+
+    try:
+        access_token = refresh_oauth_token(account)
+    except HTTPException:
+        raise HTTPException(status_code=500, detail=ErrorDetail.PROVIDER_ERROR)
+
+    try:
+        try:
+            gl = gitlab.Gitlab(url)
+            gl.groups.get(username)
+        except gitlab.GitlabError:
+            # Group does not exist. Either it's private or it's a user.
+            raise HTTPException(
+                status_code=403, detail=ErrorDetail.USERNAME_DOES_NOT_MATCH
+            )
+
+        # python-gitlab does not support the userinfo endpoint AFAICT, so we have to do it manually.
+        r = requests.get(
+            url + "/oauth/userinfo",
+            headers={"Authorization": "Bearer " + access_token},
+        )
+
+        if not r.ok:
+            raise HTTPException(status_code=500, detail=ErrorDetail.PROVIDER_ERROR)
+
+        userinfo = r.json()
+
+        # Must be owner or maintainer of the group
+        if groups := userinfo.get("https://gitlab.org/claims/groups/owner"):
```

Comment:
Not sure I get this, but will this work for the gnome/kde/freedesktop instances?